daemon/k8s_watcher.go

```diff
@@ -1735,127 +1708,109 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
+		cnpError := overallErr
 
-		// Update the status of whether the rule is enforced on this node.
-		// If we are unable to parse the CNP retrieved from the store,
-		// or if endpoints did not reach the desired policy revision
-		// after 30 seconds, then mark the rule as not being enforced.
 		if policyImportErr != nil {
-			// OK is false here because the policy wasn't imported into
-			// cilium on this node; since it wasn't imported, it also
-			// isn't enforced.
-			cnpUpdateErr = updateCNPNodeStatus(serverRuleCpy, false, false, policyImportErr, rev, serverRuleCpy.Annotations)
-		} else if ruleCopyParseErr != nil {
-			// This handles the case where the initial instance of this
-			// rule was imported into the policy repository successfully
-			// (policyImportErr == nil), but, the rule has been updated
-			// in the store soon after, and is now invalid. As such,
-			// the rule is not OK because it cannot be imported due
-			// to parsing errors, and cannot be enforced because it is
-			// not OK.
-			cnpUpdateErr = updateCNPNodeStatus(serverRuleCpy, false, false, ruleCopyParseErr, rev, serverRuleCpy.Annotations)
-		} else {
-			// If the deadline by the above context, then not all
-			// endpoints are enforcing the given policy, and
-			// waitForEpsErr will be non-nil.
-			cnpUpdateErr = updateCNPNodeStatus(serverRuleCpy, waitForEPsErr == nil, true, waitForEPsErr, rev, serverRuleCpy.Annotations)
+			cnpError = policyImportErr
 		}
 
+		updatedCNP, cnpUpdateErr = updateCNPNodeStatus(cnp, overallErr == nil, policyImportErr == nil, cnpError, rev)
 		if cnpUpdateErr == nil {
-			logger.WithField("status", serverRuleCpy.Status).Debug("successfully updated with status")
-			break
+			logger.WithField("status", updatedCNP.Status).Debug("successfully updated with status")
+			return overallErr
 		}
 
 		// Wait a small amount of time to try to update again.
 		logger.WithError(cnpUpdateErr).Warningf("update of CNP failed; sleeping for %s amount of time before trying again", updateWaitDuration)
 		time.Sleep(updateWaitDuration)
 	}
 
-	if cnpUpdateErr != nil {
-		return cnpUpdateErr
-	} else {
-		return waitForEPsErr
+	// If the process of waiting for endpoints has been successful but the
+	// CNP status report failed, report an error to cause the controller to
+	// retry.
+	if overallErr != nil && cnpUpdateErr != nil {
+		overallErr = cnpUpdateErr
 	}
+
+	return overallErr
 }
 
-func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, err error, rev uint64, annotations map[string]string) error {
-	var (
-		cnpns cilium_v2.CiliumNetworkPolicyNodeStatus
-		err2  error
-	)
+func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, importError error, rev uint64) (*cilium_v2.CiliumNetworkPolicy, error) {
+	ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
 
+	// Retrieve the latest revision from the server to minimize the risk of
+	// a locally corrupted CNP makes its way back into the apiserver
```

Comment:
@tgraf how can a policy be locally corrupted? What difference does it make if the policy is in the cache or we are fetching from kube-apiserver? By retrieving it from kube-apiserver there's no guarantees that we don't corrupt it while changing its status and update the corrupted policy to the kube-apiserver. If you are worried that we somehow forget to do a `DeepCopy` it's just a matter of adding a unit test and check the correctness of a policy that had its status modified while it was being added / modified / delete by a k8s event.

Reply:
It's less about guarantees and more about the principle of exposed scope. While it is in the store, any code in the entire agent could potentially corrupt it. It's simply a matter of using a agent wide cache to which all code has access to or logic that is local to a particular function. I did not find any evidence that the CNP is being corrupted but if it is, the side effects on cluster level are horrendous. It is very simple to validate the correctness of a function. It is very hard to guarantee the correctness of the overall codebase long term.

I'm not worried about the current code. I'm worried about code that we will add in the future. With this change, this fragile logic does not have any external dependencies and the concept of the cache can even change without affecting this.